### ISSUE TITLE
[Bug]: builtins.ModuleNotFoundError: No module named 'attrs'

### ISSUE THREAD

**Issue by j75** [cnumr/ecoindex_cli#231]
### What happened?

Crash and no report generated, not even a log file!

The last lines displayed on my console are shown in the "Relevant log output" section below.

I am on Ubuntu 22.04 running Python 3.10.6

### Version

Above 3.6

### What OS do you use?

Linux

### urls

https://server.famion.eu/

### Relevant log output

```shell
File "<frozen importlib._bootstrap_external>", line 883, in exec_module
    
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
    
  File "/home/my/.local/lib/python3.10/site-packages/scrapy/downloadermiddlewares/retry.py", line 25, in <module>
    from twisted.web.client import ResponseFailed
  File "/home/my/.local/lib/python3.10/site-packages/twisted/web/client.py", line 24, in <module>
    from twisted.internet.endpoints import HostnameEndpoint, wrapClientTLS
  File "/home/my/.local/lib/python3.10/site-packages/twisted/internet/endpoints.py", line 63, in <module>
    from twisted.python.systemd import ListenFDs
  File "/home/my/.local/lib/python3.10/site-packages/twisted/python/systemd.py", line 18, in <module>
    from attrs import Factory, define
builtins.ModuleNotFoundError: No module named 'attrs'
```

### Code of Conduct

- [ ] I agree to follow this project's Code of Conduct

**Comment by vvatelot** [comment]
Hello @j75 , thanks for this report.
Could you tell me a little more please ?

How did you install ecoindex_cli? (pip global or poetry for dev ?)

If you used pip, what does display this command ?
```
pip show ecoindex ecoindex-scraper ecoindex-cli 
```

Or if you used poetry, what does this command display ?
```
poetry show | grep ecoindex
```

**Comment by j75** [comment]
I used **pip**. Here is the result demanded:
```
$ pip show ecoindex ecoindex-scraper ecoindex-cli 
Name: ecoindex
Version: 5.4.1
Summary: Ecoindex module provides a simple way to measure the Ecoindex score based on the 3 parameters: The DOM elements of the page, the size of the page and the number of external requests of the page
Home-page: http://www.ecoindex.fr
Author: Vincent Vatelot
Author-email: vincent.vatelot@ik.me
License: MIT
Location: /home/my/.local/lib/python3.10/site-packages
Requires: pydantic
Required-by: ecoindex-scraper
---
Name: ecoindex-scraper
Version: 2.14.0
Summary: Ecoindex_scraper module provides a way to scrape data from given website while simulating a real web browser
Home-page: http://www.ecoindex.fr
Author: Vincent Vatelot
Author-email: vincent.vatelot@ik.me
License: MIT
Location: /home/my/.local/lib/python3.10/site-packages
Requires: ecoindex, Pillow, undetected-chromedriver
Required-by: ecoindex-cli
---
Name: ecoindex-cli
Version: 2.16.1
Summary: `ecoindex-cli` is a CLI tool that let you make ecoindex tests on given pages
Home-page: https://www.ecoindex.fr
Author: Vincent Vatelot
Author-email: vincent.vatelot@ik.me
License: MIT
Location: /home/my/.local/lib/python3.10/site-packages
Requires: click-spinner, ecoindex-scraper, Jinja2, matplotlib, pandas, pydantic, PyYAML, Scrapy, typer
Required-by: 
```

**Comment by j75** [comment]
I run again the command, the result is now different!
```
/home/my/.local/bin/ecoindex-cli analyze --url https://server.famion.eu/ --output-file /tmp/ecoindex.csv
📁️ Urls recorded in file `/tmp/ecoindex-cli/input/server.famion.eu.csv`
There are 1 url(s), do you want to process? [Y/n]: y
1 urls for 1 window size with 4 maximum workers
Processing  [####################################]  1/1  100%
Errors found: please look at /tmp/ecoindex-cli/logs/server.famion.eu.log)

ExportFormat.csv
Traceback (most recent call last):

  File "/home/my/.local/bin/ecoindex-cli", line 8, in <module>
    sys.exit(app())

  File "/home/my/.local/lib/python3.10/site-packages/ecoindex_cli/cli/app.py", line 195, in analyze
    write_results_to_file(

  File "/home/my/.local/lib/python3.10/site-packages/ecoindex_cli/files.py", line 69, in write_results_to_file
    file.write()

  File "/home/my/.local/lib/python3.10/site-packages/ecoindex_cli/files.py", line 37, in write
    headers = self.results[0].__dict__

IndexError: list index out of range
```

However now I get a log file (/tmp/ecoindex-cli/logs/server.famion.eu.log):
```
2023-02-12 20:58:25,363  -- https://server.famion.eu/ -- unknown error: cannot connect to chrome at 127.0.0.1:5106
1
from session not created: This version of ChromeDriver only supports Chrome version 110
Current browser version is 108.0.5359.40
```

**Comment by vvatelot** [comment]
Could you try to launch the command with the option `--chrome-version 108` ?

**Comment by j75** [comment]
```
~/.local/bin/ecoindex-cli analyze --url https://server.famion.eu/ --recursive  --chrome-version 108 

Unhandled error in Deferred:

Traceback (most recent call last):
  File "/home/my/.local/lib/python3.10/site-packages/scrapy/crawler.py", line 233, in crawl
    return self._crawl(crawler, *args, **kwargs)
  File "/home/my/.local/lib/python3.10/site-packages/scrapy/crawler.py", line 237, in _crawl
    d = crawler.crawl(*args, **kwargs)
  File "/home/my/.local/lib/python3.10/site-packages/twisted/internet/defer.py", line 1947, in unwindGenerator
    return _cancellableInlineCallbacks(gen)
  File "/home/my/.local/lib/python3.10/site-packages/twisted/internet/defer.py", line 1857, in _cancellableInlineCallbacks
    _inlineCallbacks(None, gen, status, _copy_context())
--- <exception caught here> ---
  File "/home/my/.local/lib/python3.10/site-packages/twisted/internet/defer.py", line 1697, in _inlineCallbacks
    result = context.run(gen.send, result)
  File "/home/my/.local/lib/python3.10/site-packages/scrapy/crawler.py", line 122, in crawl
    self.engine = self._create_engine()
  File "/home/my/.local/lib/python3.10/site-packages/scrapy/crawler.py", line 136, in _create_engine
    return ExecutionEngine(self, lambda _: self.stop())
  File "/home/my/.local/lib/python3.10/site-packages/scrapy/core/engine.py", line 78, in __init__
    self.downloader = downloader_cls(crawler)
  File "/home/my/.local/lib/python3.10/site-packages/scrapy/core/downloader/__init__.py", line 85, in __init__
    self.middleware = DownloaderMiddlewareManager.from_crawler(crawler)
  File "/home/my/.local/lib/python3.10/site-packages/scrapy/middleware.py", line 68, in from_crawler
    return cls.from_settings(crawler.settings, crawler)
  File "/home/my/.local/lib/python3.10/site-packages/scrapy/middleware.py", line 43, in from_settings
    mwcls = load_object(clspath)
  File "/home/my/.local/lib/python3.10/site-packages/scrapy/utils/misc.py", line 60, in load_object
    mod = import_module(module)
  File "/usr/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
    
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
    
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
    
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
    
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
    
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
    
  File "/home/my/.local/lib/python3.10/site-packages/scrapy/downloadermiddlewares/retry.py", line 25, in <module>
    from twisted.web.client import ResponseFailed
  File "/home/my/.local/lib/python3.10/site-packages/twisted/web/client.py", line 24, in <module>
    from twisted.internet.endpoints import HostnameEndpoint, wrapClientTLS
  File "/home/my/.local/lib/python3.10/site-packages/twisted/internet/endpoints.py", line 63, in <module>
    from twisted.python.systemd import ListenFDs
  File "/home/my/.local/lib/python3.10/site-packages/twisted/python/systemd.py", line 18, in <module>
    from attrs import Factory, define
builtins.ModuleNotFoundError: No module named 'attrs'
```

**Comment by vvatelot** [comment]
OK, could you try to manualy install the package `attrs` ?

```bash
pip install --user --upgrade attrs
```

**Comment by j75** [comment]
```
% pip install --user --upgrade attrs
Requirement already satisfied: attrs in /usr/lib/python3/dist-packages (21.2.0)
Collecting attrs
  Downloading attrs-22.2.0-py3-none-any.whl (60 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.0/60.0 KB 2.0 MB/s eta 0:00:00
Installing collected packages: attrs
Successfully installed attrs-22.2.0
```
And now it seems working:
```
~/.local/bin/ecoindex-cli analyze --url https://server.famion.eu/ --recursive --chrome-version 108
You are about to perform a recursive website scraping. This can take a long time. Are you sure to want to proceed? [Y/n]: y
⏲️ Crawling root url https://server.famion.eu/ -> Wait a minute!
-2023-02-13 15:07:52 [scrapy.utils.log] INFO: Scrapy 2.8.0 started (bot: scrapybot)
2023-02-13 15:07:52 [scrapy.utils.log] INFO: Versions: lxml 4.9.1.0, libxml2 2.9.14, cssselect 1.1.0, parsel 1.7.0, w3lib 2.1.1, Twisted 22.10.0, Python 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0], pyOpenSSL 21.0.0 (OpenSSL 3.0.2 15 Mar 2022), cryptography 3.4.8, Platform Linux-5.15.0-60-generic-x86_64-with-glibc2.35
2023-02-13 15:07:52 [scrapy.crawler] INFO: Overridden settings:
{'LOG_ENABLED': False}
...
🙌️ File /tmp/ecoindex-cli/output/server.famion.eu/2023-13-02_150801/results.csv written !
```
The option ` --chrome-version 108` seems to be required, otherwise I get the error `IndexError: list index out of range`.

**Comment by vvatelot** [comment]
OK, I added a tip in the readme : https://github.com/cnumr/ecoindex_cli#using-a-specific-chrome-version

If you are OK with this, I propose you to:
1. Close this issue as it seems to be fixed
2. Open another issue to improve logging (in log file / for the `IndexError: list index out of range` error)

**Comment by j75** [comment]
OK, I'll close the issue.
Please note that on Ubuntu systems where one uses Chromium and not Google Chrome (isn't it better for the environment to use a browser that not send data to Google servers in the backend? ;-)) the version is displayed with `chromium-browser --version`.

If I have several versions installed, is it possible to select the one desired by a parameter?
